### PR TITLE
feat: add packages and content packages

### LINKED ISSUE #304: Add content packages endpoint
https://github.com/rstudio/connect/pull/28446

### PR DESCRIPTION
**Pull request by tdstein**
Adds packages support.

Resolves #304 
Requires https://github.com/rstudio/connect/issues/28227

### CODE REVIEW

**Comment by tdstein** [comment]
@mmarchetti - here is a basic implementation of the '/v1/content/.../packages' endpoint.

File: src/posit/connect/packages.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    def reload(self) -> "Packages":
+        """Reload packages from the Connect server.
+
+        Returns
+        -------
+        List[Package]
+        """
+        response = self.params.session.get(self._endpoint)
+        results = response.json()
```
Comment (mmarchetti) [inline src/posit/connect/packages.py:62]: Content that is unpublished will return a 204 and empty response body to this request, which fails JSON decoding.

**Review by mmarchetti — APPROVED** [review]

File: src/posit/connect/packages.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+class Packages(Resources, Sequence[Package]):
+    """A collection of packages."""
+
+    def __init__(self, params, endpoint):
+        super().__init__(params)
+        self._endpoint = endpoint
+        self._packages = []
+        self.reload()
+
+    def __getitem__(self, index):
```
[the remark this answers is not in the document]
Comment (tdstein) [reply src/posit/connect/packages.py:32]: Shoot, I have updates that I haven't pushed, which do this. Thanks!

File: src/posit/connect/packages.py (hunk 1 above)
Comment (tdstein) [reply src/posit/connect/packages.py:62]: Thanks! I'll add it!

File: src/posit/connect/client.py [hunk 3]
```diff
@@ -269,6 +270,11 @@ def oauth(self) -> OAuth:
         """
         return OAuth(self.resource_params, self.cfg.api_key)
 
+    @property
+    @requires(version="2024.10.0-dev")
```
Comment (tdstein) [inline src/posit/connect/client.py:274]: I will bump this to 2024.11.0 after the release.

File: src/posit/connect/resources.py [hunk 4]
```diff
@@ -97,14 +110,13 @@ def __init__(self, ctx: Context, path: str, uid: str = "guid"):
...
         self._uid = uid
-        self._cache = None
 
     @abstractmethod
     def _create_instance(self, path: str, /, **kwargs: Any) -> T:
         """Create an instance of 'T'."""
         raise NotImplementedError()
 
-    def fetch(self) -> List[T]:
+    def fetch(self, **conditions) -> Iterable[T]:
```
Comment (tdstein) [inline src/posit/connect/resources.py:119]: Refactored to work on Generators as used in `GlobalPackages`.

File: src/posit/connect/resources.py [hunk 5]
```diff
@@ -114,65 +126,61 @@ def fetch(self) -> List[T]:
...
-        return len(self._data)
+        return len(list(self.fetch()))
 
     def __str__(self) -> str:
-        return str(self._data)
+        return str(list(self.fetch()))
 
     def __repr__(self) -> str:
-        return repr(self._data)
+        return repr(list(self.fetch()))
```
Comment (tdstein) [inline src/posit/connect/resources.py:183]: This could be optimized, but leaving for now.

File: src/posit/connect/context.py [hunk 6]
```diff
@@ -45,4 +45,4 @@ def version(self, value):
 
 
 class ContextManager(Protocol):
-    ctx: Context
+    _ctx: Context
```
Comment (tdstein) [inline src/posit/connect/context.py:48]: Renaming to _ctx so that it works across `Client` and `ContentPackagesMixin`

**Comment by tdstein** [comment]
I'll fix those linting errors.

File: src/posit/connect/resources.py [hunk 7]
```diff
@@ -119,65 +128,61 @@ def fetch(self) -> List[T]:
...
+                value
+                for i, value in enumerate(islice(data, start, stop))
+                if (i + start) % step == 0
+            ]
+
+        raise TypeError(f"Index must be int or slice, not {type(index).__name__}.")
 
     def __len__(self) -> int:
-        return len(self._data)
+        return len(list(self.fetch()))
```
Comment (mmarchetti) [inline src/posit/connect/resources.py:176]: Should these methods cache the result? I'm concerned about a client that writes
```
def do_something(items):
  for i in range(len(items)):
     # do something with items[i]

do_something(client.packages)
```

Comment (tdstein) [reply src/posit/connect/resources.py:176]: We've been going back and forth on caching. I actually think I should revert these changes back to using the `_data` property.

File: src/posit/connect/packages.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+    def find_by(self, **conditions):
+        return super().find_by(**conditions)
+
+
+class ContentPackagesMixin(Active):
+    """Mixin class to add a packages attribute."""
+
+    @property
+    @requires(version="2024.10.0-dev")
+    def packages(self):
```
Comment (schloerke) [inline src/posit/connect/packages.py:102]: Does this make the property disappear when the version is too low? 

Or does it present, but throws on access?

File: src/posit/connect/packages.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        name: NotRequired[str]
+        """The package name"""
+
+        version: NotRequired[str]
+        """The package version"""
+
+        hash: NotRequired[Optional[str]]
+        """Package description hash for R packages."""
+
+    @overload
```
Comment (schloerke) [inline src/posit/connect/packages.py:60]:
```suggestion
```

File: src/posit/connect/packages.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+        Returns
+        -------
+        Optional[T]
+            The first record matching the specified conditions, or `None` if no such record exists.
+        """
+
+    @overload
+    def find_by(self, **conditions): ...
+
+    def find_by(self, **conditions):
```
Comment (schloerke) [inline src/posit/connect/packages.py:93]: No need for the overload and don't lose typing support when something is misspelled or a wrong type is given

```suggestion
```

File: src/posit/connect/packages.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+        language: Required[Literal["python", "r"]]
+        """Programming language ecosystem, options are 'python' and 'r'"""
+
+        name: Required[str]
+        """The package name"""
+
+        version: Required[str]
+        """The package version"""
+
+    @overload
```
Comment (schloerke) [inline src/posit/connect/packages.py:155]: Same as above
```suggestion
```

File: src/posit/connect/packages.py [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+    def fetch(self, **conditions: Unpack[_Fetch]): ...
+
+    @overload
+    def fetch(self, **conditions): ...
+
+    def fetch(self, **conditions) -> Generator["Package"]:
```
Comment (schloerke) [inline src/posit/connect/packages.py:161]:
```suggestion
    def fetch(self, **conditions: Unpack[_Fetch]) -> Generator["Package"]:
```

File: src/posit/connect/packages.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        bundle_id: NotRequired[str]
+        """The unique identifier of the bundle this package is associated with"""
+
+        app_id: NotRequired[str]
+        """The numerical identifier of the application this package is associated with"""
+
+        app_guid: NotRequired[str]
+        """The unique identifier of the application this package is associated with"""
+
+    @overload
```
Comment (schloerke) [inline src/posit/connect/packages.py:197]:
```suggestion
```

File: src/posit/connect/packages.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        Returns
+        -------
+        Optional[Package]
+            The first record matching the specified conditions, or `None` if no such record exists.
+        """
+
+    @overload
+    def find_by(self, **conditions) -> "Package | None": ...
+
+    def find_by(self, **conditions) -> "Package | None":
```
Comment (schloerke) [inline src/posit/connect/packages.py:230]: Same as above
```suggestion
```

**Review by schloerke — CHANGES_REQUESTED** [review]

File: src/posit/connect/packages.py (hunk 8 above)
Comment (tdstein) [reply src/posit/connect/packages.py:102]: It's available, but throws a NotImplementedError on access.

Comment (schloerke) [reply src/posit/connect/packages.py:102]: Cool!

File: src/posit/connect/packages.py (hunk 10 above)
Comment (tdstein) [reply src/posit/connect/packages.py:93]: I'm seeing...

```
Method "fetch" overrides class "ActiveSequence" in an incompatible manner
  Parameter "**conditions" has no corresponding parameterPylance[reportIncompatibleMethodOverride](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportIncompatibleMethodOverride)
```

I'm looking for a workaround, but no luck so far.

**Review by schloerke — APPROVED** [review]

File: src/posit/connect/resources.py [hunk 15]
```diff
@@ -119,65 +128,61 @@ def fetch(self) -> List[T]:
...
+        data = self.fetch()
+
+        if isinstance(index, int):
+            if index < 0:
+                # Handle negative indexing
+                data = list(data)
+                return data[index]
+            for i, value in enumerate(data):
+                if i == index:
+                    return value
```
Comment (toph-allen) [inline src/posit/connect/resources.py]: I'm guessing that enumeration is cheaper than just making whatever data into a list?

**Review by toph-allen — COMMENTED** [review]
This review was half-done when the LLM hackathon kickoff started and now my pending comment is out of date and not displayed. Oops.

File: src/posit/connect/resources.py (hunk 15 above)
Comment (tdstein) [reply src/posit/connect/resources.py]: No worries! 😛 Hope the LLM kickoff went well.

That's correct. The enumerate operates on the iterable, which avoids pulling the entire list into memory.

**Pull request merged**